src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -175,14 +176,60 @@
                                      (= (.toPath ^File (:changed-path event)) (:file callback-entry)))))]
     (filter filter-fn events)))
 
+(defn ->tuple
+  "Create a tuple of the two fields we care about in an Event.
+  This simple element querying our sets simple."
+  [event]
+  [(:changed-path event) (:type event)])
+
+(defn already-seen?
+  "Expects a set of event tuples (see ->tuple) and a full Event.
+  Is this a modify event for a changed-path where we've already seen a create or modify event?"
+  [events {:keys [changed-path type] :as _event}]
+  (when (= :modify type)
+     (or (set/subset? #{[changed-path :create]} events)
+         (set/subset? #{[changed-path :modify]} events))))
+
+(defn should-overwrite?
+  "Expects a set of event tuples (see ->tuple) and a full Event.
+  Is this a create event that should supersede a previously found modify or create event?"
+  [events {:keys [changed-path type] :as _event}]
+  (when (= :create type)
+     (or (set/subset? #{[changed-path :modify]} events)
+         (set/subset? #{[changed-path :create]} events))))
+
+(defn reducible-remove
+  [{:keys [tuples] :as memo} item]
+  (cond
+    ;; Ignore the new item if we've already seen a value representing it
+    (already-seen? tuples item)
+      memo
+    ;; Remove old value and replace it with the new value
+    (should-overwrite? tuples item)
+      (-> memo
+          (update :tuples set/difference #{[(:changed-path item) :modify]})
+          (update :tuples set/union #{(->tuple item)})
+          (update :events (fn [es] (filter #(and (= (:changed-path item) (:changed-path %))
+                                                (or
+                                                  (= :modify (:type %))
+                                                  (= :create (:type %))))
+                                           es)))
+          (update :events conj item))
+    ;; default add the event to our tracker
+    :else
+      (-> memo
+          (update :tuples set/union #{(->tuple item)})
+          (update :events conj item))))
+
 (schema/defn process-events!
   "Process for side effects any events that occurred for watcher's watch-key"
   [watcher :- (schema/protocol Watcher)
    events :- [Event]]
-  (let [callbacks @(:callbacks watcher)]
+  (let [callbacks @(:callbacks watcher)
+        compressed-events (->> events (reduce reducible-remove {:tuples #{} :events []}) :events)]
```

Comment:
Would all `events` have the same `changed-path`? In other words, is a collection of events here for a single file?

Reply:
Answer is yes! We need to manage the compression per changed path.